beacon-chain/rpc/eth/beacon/handlers_validator.go

```diff
@@ -0,0 +1,375 @@
+
+	rawIds := r.URL.Query()["id"]
+	ids, ok := decodeIds(w, st, rawIds, true /* ignore unknown */)
+	if !ok {
+		return
+	}
+	// return no data if all IDs are ignored
+	if len(rawIds) > 0 && len(ids) == 0 {
+		resp := &GetValidatorBalancesResponse{
+			Data:                []*ValidatorBalance{},
+			ExecutionOptimistic: isOptimistic,
+			Finalized:           isFinalized,
+		}
+		http2.WriteJson(w, resp)
+		return
+	}
+
+	bals := st.Balances()
+	var valBalances []*ValidatorBalance
+	if len(ids) == 0 {
+		valBalances = make([]*ValidatorBalance, len(bals))
+		for i, b := range bals {
+			valBalances[i] = &ValidatorBalance{
+				Index:   strconv.FormatUint(uint64(i), 10),
+				Balance: strconv.FormatUint(b, 10),
+			}
+		}
+	} else {
+		valBalances = make([]*ValidatorBalance, len(ids))
+		for i, id := range ids {
+			valBalances[i] = &ValidatorBalance{
+				Index:   strconv.FormatUint(uint64(id), 10),
+				Balance: strconv.FormatUint(bals[id], 10),
+			}
+		}
+	}
+
+	resp := &GetValidatorBalancesResponse{
+		Data:                valBalances,
+		ExecutionOptimistic: isOptimistic,
+		Finalized:           isFinalized,
+	}
+	http2.WriteJson(w, resp)
+}
+
+// decodeIds takes in a list of validator ID strings (as either a pubkey or a validator index)
+// and returns the corresponding validator indices. It can be configured to ignore well-formed but unknown indices.
+func decodeIds(w http.ResponseWriter, st state.BeaconState, rawIds []string, ignoreUnknown bool) ([]primitives.ValidatorIndex, bool) {
+	ids := make([]primitives.ValidatorIndex, 0, len(rawIds))
+	numVals := uint64(st.NumValidators())
+	for _, rawId := range rawIds {
+		pubkey, err := hexutil.Decode(rawId)
+		if err == nil {
+			if len(pubkey) != fieldparams.BLSPubkeyLength {
+				http2.HandleError(w, fmt.Sprintf("Pubkey length is %d instead of %d", len(pubkey), fieldparams.BLSPubkeyLength), http.StatusBadRequest)
+				return nil, false
+			}
+			valIndex, ok := st.ValidatorIndexByPubkey(bytesutil.ToBytes48(pubkey))
+			if !ok {
+				if ignoreUnknown {
+					continue
+				}
+				http2.HandleError(w, fmt.Sprintf("Unknown pubkey %s", pubkey), http.StatusBadRequest)
+				return nil, false
+			}
+			ids = append(ids, valIndex)
+			continue
+		}
+
+		index, err := strconv.ParseUint(rawId, 10, 64)
+		if err != nil {
+			http2.HandleError(w, fmt.Sprintf("Invalid validator index %s", rawId), http.StatusBadRequest)
+			return nil, false
+		}
+		if index >= numVals {
+			if ignoreUnknown {
+				continue
+			}
+			http2.HandleError(w, fmt.Sprintf("Invalid validator index %d", index), http.StatusBadRequest)
+			return nil, false
+		}
+		ids = append(ids, primitives.ValidatorIndex(index))
+	}
+	return ids, true
+}
+
+// valsFromIds returns read-only validators based on the supplied validator indices.
+func valsFromIds(w http.ResponseWriter, st state.BeaconState, ids []primitives.ValidatorIndex) ([]state.ReadOnlyValidator, bool) {
+	var vals []state.ReadOnlyValidator
+	if len(ids) == 0 {
```

Comment:
I don't know when this ever gets triggered or should be triggered as ids can't be 0 from GetValidator or GetValidators functions

Reply:
It can be 0 in GetValidators in two cases:
- no validator IDs are passed
- all passed IDs are ignored

Reply:
but when it's ignored you'll exit early 
```
if len(rawIds) > 0 && len(ids) == 0 {
		resp := &GetValidatorsResponse{
			Data:                []*ValidatorContainer{},
			ExecutionOptimistic: isOptimistic,
			Finalized:           isFinalized,
		}
		http2.WriteJson(w, resp)
		return
	}
```

Reply:
You're right, but this only covers the second case. When no IDs are requested, then the `if lan(ids) == 0` condition will be triggered.

Reply:
oh decode ids can still return 0 hmm. ok